x-pack/filebeat/input/cometd/input_test.go

```diff
@@ -20,182 +21,274 @@ import (
-	logger := logp.NewLogger("test")
-	authParams := bay.AuthenticationParameters{}
-	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
-	workerCtx, workerCancel := context.WithCancel(inputCtx)
-	defer cancelInputCtx()
-
-	input := &cometdInput{
-		config:       conf,
-		log:          logger,
-		inputCtx:     inputCtx,
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-		ackedCount:   atomic.NewUint32(0),
-		authParams:   authParams,
-	}
-	input.out = make(chan bay.TriggerEvent)
+		cfg := common.MustNewConfigFrom(config)
 
-	input.Stop()
-}
+		input, err := NewInput(cfg, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input)
 
-func TestWait(t *testing.T) {
-	conf := defaultConfig()
-	logger := logp.NewLogger("test")
-	authParams := bay.AuthenticationParameters{}
-	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
-	workerCtx, workerCancel := context.WithCancel(inputCtx)
-	defer cancelInputCtx()
-
-	input := &cometdInput{
-		config:       conf,
-		log:          logger,
-		inputCtx:     inputCtx,
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-		ackedCount:   atomic.NewUint32(0),
-		authParams:   authParams,
-	}
-	input.out = make(chan bay.TriggerEvent)
+		// run input
+		input.Run()
+
+		go func() {
+			time.Sleep(100 * time.Millisecond) // let input.Stop() be executed.
+			input.Wait()
+		}()
 
-	input.Wait()
+		for range []beat.Event{<-eventsCh} {
+		}
+	})
+
+	t.Run("Check input stop", func(t *testing.T) {
+		conf := defaultConfig()
+		logger := logp.NewLogger("test")
+		authParams := bay.AuthenticationParameters{}
+		inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+		workerCtx, workerCancel := context.WithCancel(inputCtx)
+		defer cancelInputCtx()
+
+		input := &cometdInput{
+			config:       conf,
+			log:          logger,
+			inputCtx:     inputCtx,
+			workerCtx:    workerCtx,
+			workerCancel: workerCancel,
+			authParams:   authParams,
+		}
+		input.msgCh = make(chan bay.MaybeMsg)
+
+		input.Stop()
+		select {
+		case <-workerCtx.Done():
+		default:
+			require.NoError(t, fmt.Errorf("input is not stopped."))
+		}
+	})
+
+	t.Run("Check input wait", func(t *testing.T) {
+		conf := defaultConfig()
+		logger := logp.NewLogger("test")
+		authParams := bay.AuthenticationParameters{}
+		inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+		workerCtx, workerCancel := context.WithCancel(inputCtx)
+		defer cancelInputCtx()
+
+		input := &cometdInput{
+			config:       conf,
+			log:          logger,
+			inputCtx:     inputCtx,
+			workerCtx:    workerCtx,
+			workerCancel: workerCancel,
+			authParams:   authParams,
+		}
+		input.msgCh = make(chan bay.MaybeMsg)
+
+		go func() {
+			time.Sleep(1000 * time.Millisecond)
+			input.Wait()
+		}()
+
+		time.Sleep(1000 * time.Millisecond) // let input.Stop() be executed.
+		select {
+		case <-workerCtx.Done():
+		default:
+			require.NoError(t, fmt.Errorf("input is not stopped."))
+		}
```

Comment:
```suggestion
		go func() {
			input.Wait()
		}()

		select {
		case <-workerCtx.Done():
		case <-time.After(time.Second): // let input.Stop() be executed.
			require.NoError(t, fmt.Errorf("input is not stopped."))
		}
```

You should be able to simplify this to eliminate the need for a delay.